### ISSUE TITLE
zstd: all the goroutines stuck in blockDec.startDecoder in v1.9.4

### ISSUE THREAD

**Issue by valyala** [klauspost/compress#195]
After upgrading from v1.9.2 to v1.9.4 in [VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics) we see a lot of goroutines stuck (i.e. 12K goroutines and counting - looks like a leak) on the line 195 of [blockDec.startDecoder](https://github.com/klauspost/compress/blob/438db640fd672278e44bb9bec0b979358e7e9c76/zstd/blockdec.go#L195).

**Comment by klauspost** [comment]
Could you direct me to the place where it is used?

**Comment by klauspost** [comment]
Encoders are kept around until Close is called.  See https://github.com/klauspost/compress/tree/master/zstd#allocation-less-operation

**Comment by valyala** [comment]
> Could you direct me to the place where it is used?

See https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/encoding/zstd/zstd_pure.go . This code didn't change for the last 5 months and it was working without issues until the upgrade from v1.9.2 to v1.9.2 .

**Comment by klauspost** [comment]
There will be gouroutines running, since they are spun up when you create the decoder. By default it should be the same as your GOMAXPROCS. This hasn't changed and I am a bit skeptical since tests actually check that no gouroutines leak.

You encoder has a "race", though. If other gouroutines are blocked by your mutex they will create a new encoder, leaking the existing, here: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/encoding/zstd/zstd_pure.go#L56

The easiest solution is to recheck when you actually get the lock whether someone else has creates it in the meantime.

**Comment by valyala** [comment]
> There will be gouroutines running, since they are spun up when you create the decoder. By default it should be the same as your GOMAXPROCS. This hasn't changed and I am a bit skeptical since tests actually check that no gouroutines leak.

GOMAXPROCS goroutines look correct. But we have 12K goroutines stuck at the line 195 of [blockDec.startDecoder](https://github.com/klauspost/compress/blob/438db640fd672278e44bb9bec0b979358e7e9c76/zstd/blockdec.go#L195) and this number constantly grows after the upgrade from v1.9.2 to v1.9.4 .

> You encoder has a "race", though. If other gouroutines are blocked by your mutex they will create a new encoder, leaking the existing, here: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/encoding/zstd/zstd_pure.go#L56

Thanks for pointing to the issue with encoder. It will be fixed soon. Note that the original issue is in the decoder, not encoder.

**Comment by klauspost** [comment]
Yeah, 12K definitely shouldn't happen. I will take a closer look in a few days, holidays at the moment.

**Comment by valyala** [comment]
Thanks!

The issue with the encoder should be fixed in the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/6a1499efa3e5980e2d83eab86a64033aeb3f1be8 .

**Comment by valyala** [comment]
It looks like I found the original issue in the cluster branch, which could lead to goroutines leak - see https://github.com/VictoriaMetrics/VictoriaMetrics/commit/108a60d69e2e74f8f375ea60d91f813eef41c546 (the typo has been fixed in the next commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/44f886cc9c5b01b868ddf374e7c9dc23dee0a684 .

It looks like the goroutine leak was here for the last 5 months, but it wasn't noticed earlier because compressed connections in cluster version of VictoriaMetrics usually live for long time. Probably something has been changed in VictoriaMetrics code at the same time as the upgrade from v1.9.2 to v1.9.4 for `github.com/klauspost/compress` library, which led to increased connection churn and goroutines leak. This needs to be investigated on VictoriaMetrics side.

Closing this issue as false alarm.
